### PR TITLE
tickets: cache ti.to data & load fresh via AJAX

### PR DESCRIPTION
**Pull request by messa**
Stránka tickets se načte hned, ale _možná_ na ni budou trochu starší nakešované údaje. Hned po načtení stránky se spustí načtení toho hlavního HTML obsahu znovu (`$(...).load(...)`) přes ajax. Ti.to widgety se aktivují až po tom ajaxovém načtení, ty buttony jsou do té doby disabled.

Akorát to bude chtít pořádně otestovat, protože když to bude bibnout... tak si lidi nekoupí lístky :/

V tom jQuery slim zřejmě nebyla podpora AJAXu, tak jsem to upgradnul na plnotučné jQuery. Sorry :)

Ten CSS loader jsem nějak vygooglil. Nemá @benabraham něco lepšího? :)

Screenshot té načítací fáze:

image

### CODE REVIEW

**Comment by messa** [comment]
Trochu off-topic: Tyhle přístupy na další API jsou v klasických webových aplikacích (Django, Flask...) prostě nic moc. Zaměstnává to workery, kterých je pak málo a snadno se udělá DoS (i neúmyslný). Stačí si stránku tickets otevřít v pár tabech (alternativa: hodí se info na Twitter a pár lidí na to klikne) a web pyconu nejede. Dá se nakonfigurovat, aby běželo víc workerů, jenže zase pokud se něco kešuje v paměti workeru, tak pak má webovka tucty izolovaných keší. Ideální je na to async přístup. Dokonce lze pak snadno dělat to, že víc workerů si počká na dokončení jednoho API requestu. V Djangu se async asi nějak dá dělat přes channels. Bohužel netuším, jak se to nastavuje.

**Comment by starenka** [comment]
Nebo treba na ten api call dat timeout ;)

**Comment by frenzymadness** [comment]
Zkusil jsem to a funguje to. Nicméně se na to bude muset podívat zkušenějším JS okem @benabraham a to bude nejdříve zítra. Mimo to bude potřeba přegenerovat styly (sass) a nějakou další mašinérii.

**Comment by benabraham** [comment]
Máme to postavený na Bootstrapu a ten má spinner v sobě: https://getbootstrap.com/docs/4.3/components/spinners/

**Comment by messa** [comment]
@benabraham Díky za info, použil jsem ho.

U toho tito buttonu jsem snížil padding, aby se text pokud možno nezalamoval. Ale zase aby se nezmenšilo celé tlačítko, tak jsem přidal min-width 80%.
Nastavil jsem .pc-ticket min-width, aby se nestávalo, že vedle sebe budou namačkané 4 ty karty (nebo jak se to jmenuje), když by mohly být už zalomené.
Oba tyto problémy se stávají jen v určitých šířkách prohlížeče. Ale nejsem CSS guru, tak snad jsem naopak žádnou situaci nezhoršil.

**Comment by hroncok** [comment]
BTW na async django je proposal na pycon ;)

**Comment by messa** [comment]
Na asyncio graphql web api taky 😎

so 9. 3. 2019 v 11:56 odesílatel Miro Hrončok <notifications@github.com>
napsal:

> BTW na async django je proposal na pycon ;)
>

**Comment by hroncok** [comment]
Můžeme to dát na betu?

**Comment by frenzymadness** [comment]
Je to na betě

**Comment by hroncok** [comment]
když se dokroutí loading je tam takový hnusný probliknutí než se objeví tlačítko buy ticket

**Comment by hroncok** [comment]
Bez JS to píše loading do nekonečna místo toho, aby tam byl odkaz na nákup jako předtím.

**Comment by messa** [comment]
> Bez JS to píše loading do nekonečna místo toho, aby tam byl odkaz na nákup jako předtím.

Opravdu to tam bylo? A dá se na to kliknout? Koupit lístek? :)

**Comment by hroncok** [comment]
jasně, otevře to tu stránku na titu, je to normální blbej odkaz.

**Comment by messa** [comment]
> když se dokroutí loading je tam takový hnusný probliknutí než se objeví tlačítko buy ticket

Njn. Není to React. A míra hnustosti toho probliknutí nejspíš záleží na prohlížeči.

Na druhou stranu, aspoň si uživatel všimne, že se něco stalo :)

Beru to tak, že jde o zlepšení situace. Nebo minimálně nezhoršení. Předtímta stránka byla perfektní, ale objevila se až za 7+ sekund, což je někdy i fatální (v ecommerce).

**Comment by messa** [comment]
> jasně, otevře to tu stránku na titu, je to normální blbej odkaz.

Zkoušel jsi to?

**Comment by messa** [comment]
Já když si vypnu Javascript a otevřu https://cz.pycon.org/2019/tickets/ (kde je teď ještě původní verze), tak to není odkaz, ale nefunkční nápis Buy ticket. Protože odkaz z toho udělá... až javascript načítaný z tito.

**Comment by messa** [comment]
Možná tam ten odkaz byl dřív: https://github.com/pyvec/cz.pycon.org-2019/blob/master/pyconcz/templates/default/pages/_ticket-button.html

Asi bych to mohl upravit, aby už v té nakešované verzi byl noscript s odkazem na tito.

**Comment by hroncok** [comment]
Myslím, že jsem to zkoušel, ale je fakt, že nevím v jaké fázi

**Comment by benabraham** [comment]
Fallback s odkazem jsme odstranili.

**Comment by messa** [comment]
> Fallback s odkazem jsme odstranili.

@benabraham Mám ho tam vrátit do nějakého `<noscript>`?

**Comment by benabraham** [comment]
otázka je, jestli ti.to vůbec bez js funguje (nemůžu otestovat teď), pokud ano, tak by bylo fajn tam mít odkaz

**Comment by messa** [comment]
@benabraham Přidal jsem podporu pro vypnutý Javascript.

**Comment by hroncok** [comment]
Počkal bych ještě, jak se to chová s vyprodaným early birdem (to se stane za chvíli). Ale jinak to podle mě funguje dobře.

**Comment by hroncok** [comment]
za mě asi dobrý. díky
